File: js/src/cors.js
/**
 * Cross-origin resource sharing (specification §8).
 */

import { appendVary } from "./headers.js";

/** Headers a browser must be allowed to send to negotiate Links Notation. */
export const DEFAULT_ALLOWED_HEADERS = [
  "Content-Type",
  "Accept",
  "Authorization",
  "If-Match",
  "If-None-Match",
];

/** Headers a browser must be allowed to read to follow the protocol. */
export const DEFAULT_EXPOSED_HEADERS = ["ETag", "Link", "Location", "Allow"];

const DEFAULT_METHODS = [
  "GET",
  "HEAD",
  "OPTIONS",
  "POST",
  "PUT",
  "PATCH",
  "DELETE",
];

/**
 * Build the CORS response headers for a request.
 *
 * @param {object} [options] - Policy
 * @param {string|string[]} [options.origin] - Allowed origin(s), `"*"` by default
 * @param {string[]} [options.methods] - Allowed methods
 * @param {string[]} [options.allowedHeaders] - Allowed request headers
 * @param {string[]} [options.exposedHeaders] - Response headers exposed to scripts
 * @param {boolean} [options.credentials] - Allow credentialed requests
 * @param {number} [options.maxAge] - Preflight cache lifetime in seconds
 * @param {string} [requestOrigin] - `Origin` header of the request
 * @returns {object} Response headers
 */
export function corsHeaders(options = {}, requestOrigin) {
  const {
    origin = "*",
    methods = DEFAULT_METHODS,
    allowedHeaders = DEFAULT_ALLOWED_HEADERS,
    exposedHeaders = DEFAULT_EXPOSED_HEADERS,
    credentials = false,
    maxAge = 600,
  } = options;

  const allowedOrigins = Array.isArray(origin) ? origin : [origin];
  let allowOrigin = null;
  if (allowedOrigins.includes("*")) {
    allowOrigin = credentials && requestOrigin ? requestOrigin : "*";
  } else if (requestOrigin && allowedOrigins.includes(requestOrigin)) {
    allowOrigin = requestOrigin;
  }

  if (!allowOrigin) {
    return {};
  }

  const headers = {
    "Access-Control-Allow-Origin": allowOrigin,
    "Access-Control-Allow-Methods": methods.join(", "),
    "Access-Control-Allow-Headers": allowedHeaders.join(", "),
    "Access-Control-Expose-Headers": exposedHeaders.join(", "),
    "Access-Control-Max-Age": String(maxAge),
  };
  if (credentials) {
    headers["Access-Control-Allow-Credentials"] = "true";
  }
  return headers;
}

/**
 * Express middleware applying a CORS policy and answering preflight requests.
 *
 * @param {object} [options] - Policy, see {@link corsHeaders}
 * @returns {Function} Express middleware
 */
export function linoCors(options = {}) {
  return (req, res, next) => {
    const headers = corsHeaders(options, req.headers.origin);
    for (const [name, value] of Object.entries(headers)) {
      res.set(name, value);
    }
    if (headers["Access-Control-Allow-Origin"] !== "*") {
      appendVary(res, "Origin");
    }
    if (
      req.method === "OPTIONS" &&
      req.headers["access-control-request-method"]
    ) {
      res.status(204).end();
      return;
    }
    next();
  };
}

File: rust/src/cors.rs
//! Cross-origin resource sharing (specification section 8).

use crate::headers::Headers;

/// Headers a browser must be allowed to send to negotiate Links Notation.
pub const DEFAULT_ALLOWED_HEADERS: [&str; 5] = [
    "Content-Type",
    "Accept",
    "Authorization",
    "If-Match",
    "If-None-Match",
];

/// Headers a browser must be allowed to read to follow the protocol.
pub const DEFAULT_EXPOSED_HEADERS: [&str; 4] = ["ETag", "Link", "Location", "Allow"];

/// Methods a browser may use unless the policy narrows them.
pub const DEFAULT_METHODS: [&str; 7] = ["GET", "HEAD", "OPTIONS", "POST", "PUT", "PATCH", "DELETE"];

/// A cross-origin policy.
///
/// [`CorsPolicy::permissive`] is the `cors: true` of the JavaScript and Python
/// packages: every origin, every default method and header.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CorsPolicy {
    /// Allowed origins; a single `*` allows every origin.
    pub origins: Vec<String>,
    /// Allowed methods.
    pub methods: Vec<String>,
    /// Headers a request may carry.
    pub allowed_headers: Vec<String>,
    /// Headers a browser may read from the response.
    pub exposed_headers: Vec<String>,
    /// Whether credentials may be sent.
    pub credentials: bool,
    /// How long a preflight result may be cached, in seconds.
    pub max_age: u32,
}

impl Default for CorsPolicy {
    fn default() -> Self {
        Self::permissive()
    }
}

impl CorsPolicy {
    /// The permissive default policy.
    pub fn permissive() -> Self {
        Self {
            origins: vec!["*".to_string()],
            methods: DEFAULT_METHODS.iter().map(|m| m.to_string()).collect(),
            allowed_headers: DEFAULT_ALLOWED_HEADERS
                .iter()
                .map(|h| h.to_string())
                .collect(),
            exposed_headers: DEFAULT_EXPOSED_HEADERS
                .iter()
                .map(|h| h.to_string())
                .collect(),
            credentials: false,
            max_age: 600,
        }
    }

    /// The same policy restricted to the given origins.
    ///
    /// # Examples
    ///
    /// ```
    /// use lino_rest_api::cors::{cors_headers, CorsPolicy};
    ///
    /// let policy = CorsPolicy::permissive().with_origins(["https://app.example"]);
    /// assert!(cors_headers(&policy, Some("https://other.example")).is_empty());
    /// ```
    pub fn with_origins<I, S>(mut self, origins: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.origins = origins.into_iter().map(Into::into).collect();
        self
    }

    /// The same policy with credentials allowed.
    pub fn with_credentials(mut self, credentials: bool) -> Self {
        self.credentials = credentials;
        self
    }

    /// The same policy with another preflight cache lifetime.
    pub fn with_max_age(mut self, max_age: u32) -> Self {
        self.max_age = max_age;
        self
    }
}

/// Build the CORS response headers for a request.
///
/// The map is empty when the origin is not allowed, which leaves the response
/// without any cross-origin permission — exactly what a browser needs to refuse
/// it.
pub fn cors_headers(policy: &CorsPolicy, request_origin: Option<&str>) -> Headers {
    let wildcard = policy.origins.iter().any(|origin| origin == "*");
    let allow_origin = if wildcard {
        match (policy.credentials, request_origin) {
            (true, Some(origin)) => Some(origin.to_string()),
            _ => Some("*".to_string()),
        }
    } else {
        request_origin
            .filter(|origin| policy.origins.iter().any(|allowed| allowed == origin))
            .map(str::to_string)
    };

    let Some(allow_origin) = allow_origin else {
        return Headers::new();
    };

    let mut headers = Headers::from_pairs([
        ("Access-Control-Allow-Origin", allow_origin),
        ("Access-Control-Allow-Methods", policy.methods.join(", ")),
        (
            "Access-Control-Allow-Headers",
            policy.allowed_headers.join(", "),
        ),
        (
            "Access-Control-Expose-Headers",
            policy.exposed_headers.join(", "),
        ),
        ("Access-Control-Max-Age", policy.max_age.to_string()),
    ]);
    if policy.credentials {
        headers.insert("Access-Control-Allow-Credentials", "true");
    }
    headers
}
